In [1]:
import pandas as pd
import numpy as np
import os

In [2]:
col_names = ['loadingOrder','carrierName','timestamp','longitude',
                  'latitude','vesselMMSI','speed','direction','vesselNextport',
                  'vesselNextportETA','vesselStatus','vesselDatasource','TRANSPORT_TRACE']
df = pd.read_csv("data/train0523.csv",header =None, names = col_names, iterator=True)

In [3]:
df_test = pd.read_csv("data/A_testData0531.csv")

In [4]:
df_sample = df.get_chunk(100000)
df_sample.shape

(100000, 13)

In [5]:
df_sample.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 13 columns):
 #   Column             Non-Null Count   Dtype  
---  ------             --------------   -----  
 0   loadingOrder       100000 non-null  object 
 1   carrierName        100000 non-null  object 
 2   timestamp          100000 non-null  object 
 3   longitude          100000 non-null  float64
 4   latitude           100000 non-null  float64
 5   vesselMMSI         100000 non-null  object 
 6   speed              100000 non-null  int64  
 7   direction          100000 non-null  int64  
 8   vesselNextport     71721 non-null   object 
 9   vesselNextportETA  71721 non-null   object 
 10  vesselStatus       71721 non-null   object 
 11  vesselDatasource   100000 non-null  object 
 12  TRANSPORT_TRACE    1554 non-null    object 
dtypes: float64(2), int64(2), object(9)
memory usage: 9.9+ MB


In [ ]:
df_test

In [91]:
df_test["TRANSPORT_TRACE"].unique()
df_test["Start"] = df_test["TRANSPORT_TRACE"].apply(lambda x: x.split("-")[0])
df_test["End"] = df_test["TRANSPORT_TRACE"].apply(lambda x: x.split("-")[-1])
df_test

loadingOrder                 timestamp   longitude   latitude  speed  \
0      CF946210847851  2019-04-02T02:42:28.000Z  138.471062  40.278787     31   
1      CF946210847851  2019-04-02T02:59:28.000Z  138.552168  40.327785     30   
2      CF946210847851  2019-04-02T03:07:28.000Z  138.588250  40.352542     30   
3      CF946210847851  2019-04-02T03:43:28.000Z  138.751325  40.459447     30   
4      CF946210847851  2019-04-02T04:29:28.000Z  138.969782  40.581485     30   
...               ...                       ...         ...        ...    ...   
45451  XG479584941731  2019-01-13T03:56:08.000Z  104.633357   1.630708      0   
45452  XG479584941731  2019-01-13T03:57:08.000Z  104.631958   1.626713      0   
45453  XG479584941731  2019-01-13T03:57:38.000Z  104.631258   1.624615      0   
45454  XG479584941731  2019-01-13T03:58:08.000Z  104.630597   1.622682      0   
45455  XG479584941731  2019-01-13T03:59:08.000Z  104.629178   1.618552      0   

       direction carrierName   vesselMMSI           onboardDate  \
0           5800      OIEQNT  R5480015614  2019/04/02  02:42:28   
1           4600      OIEQNT  R5480015614  2019/04/02  02:42:28   
2           4900      OIEQNT  R5480015614  2019/04/02  02:42:28   
3           5000      OIEQNT  R5480015614  2019/04/02  02:42:28   
4           5000      OIEQNT  R5480015614  2019/04/02  02:42:28   
...          ...         ...          ...                   ...   
45451      19890      JCMFTA  U2218600548  2019/01/10  00:27:58   
45452      19890      JCMFTA  U2218600548  2019/01/10  00:27:58   
45453      19900      JCMFTA  U2218600548  2019/01/10  00:27:58   
45454      19910      JCMFTA  U2218600548  2019/01/10  00:27:58   
45455      19930      JCMFTA  U2218600548  2019/01/10  00:27:58   

      TRANSPORT_TRACE  Start    End  
0         CNYTN-MXZLO  CNYTN  MXZLO  
1         CNYTN-MXZLO  CNYTN  MXZLO  
2         CNYTN-MXZLO  CNYTN  MXZLO  
3         CNYTN-MXZLO  CNYTN  MXZLO  
4         CNYTN-MXZLO  CNYTN  MXZLO  
...               ...    ...    ...  
45451     CNYTN-MATNG  CNYTN  MATNG  
45452     CNYTN-MATNG  CNYTN  MATNG  
45453     CNYTN-MATNG  CNYTN  MATNG  
45454     CNYTN-MATNG  CNYTN  MATNG  
45455     CNYTN-MATNG  CNYTN  MATNG  

[45456 rows x 12 columns]

In [7]:
df_test["Start"].unique()

array(['CNYTN', 'CNSHK', 'COBUN', 'CNSHA', 'CNHKG', 'HKHKG'], dtype=object)

In [8]:
df_test["End"].unique()

array(['MXZLO', 'MYTPP', 'SGSIN', 'CLVAP', 'ARENA', 'MATNG', 'GRPIR',
       'PKQCT', 'HKHKG', 'PAONX', 'SIKOP', 'CAVAN', 'ESALG', 'MTMLA',
       'ZADUR', 'LBBEY', 'RTM', 'FRFOS', 'NZAKL', 'PAMIT'], dtype=object)

## Port

In [9]:
df_sample["TRANSPORT_TRACE"].unique()

array([nan, 'CNYTN-CLIQQ', 'COLON CONTAINER TERMINAL', 'CNSHK-HRRJK',
       'SHEKOU-DAMMAM-RIYADH CITY_SAUDI ARABIA', 'CNYTN-INMAA',
       'JEBEL ALI-HONG KONG'], dtype=object)

In [10]:
df_sample.vesselStatus.unique()

array([nan, 'under way using engine', 'at anchor', 'moored',
       'not under command', 'under way sailing'], dtype=object)

In [11]:
df_sample.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 13 columns):
 #   Column             Non-Null Count   Dtype  
---  ------             --------------   -----  
 0   loadingOrder       100000 non-null  object 
 1   carrierName        100000 non-null  object 
 2   timestamp          100000 non-null  object 
 3   longitude          100000 non-null  float64
 4   latitude           100000 non-null  float64
 5   vesselMMSI         100000 non-null  object 
 6   speed              100000 non-null  int64  
 7   direction          100000 non-null  int64  
 8   vesselNextport     71721 non-null   object 
 9   vesselNextportETA  71721 non-null   object 
 10  vesselStatus       71721 non-null   object 
 11  vesselDatasource   100000 non-null  object 
 12  TRANSPORT_TRACE    1554 non-null    object 
dtypes: float64(2), int64(2), object(9)
memory usage: 9.9+ MB


比赛网站

https://competition.huaweicloud.com/information/1000037843/circumstance

In [12]:
df_sample.head()

loadingOrder carrierName                 timestamp   longitude  \
0  ZQ464072113491      FGPEPT  2019-01-01T00:00:03.000Z -106.724293   
1  ZQ464072113491      FGPEPT  2019-01-01T00:00:08.000Z -106.724880   
2  RZ321715965702      OIEQNT  2019-01-01T00:00:10.000Z  131.908943   
3  ZQ464072113491      FGPEPT  2019-01-01T00:00:14.000Z -106.725467   
4  ZQ464072113491      FGPEPT  2019-01-01T00:00:21.000Z -106.726160   

    latitude   vesselMMSI  speed  direction vesselNextport vesselNextportETA  \
0 -25.831867  G8140282065     37      29020            NaN               NaN   
1 -25.831680  G8140282065     37      28960            NaN               NaN   
2  31.147950  J1826969247     32       7600            NaN               NaN   
3 -25.831493  G8140282065     37      29010            NaN               NaN   
4 -25.831280  G8140282065     37      28940            NaN               NaN   

  vesselStatus vesselDatasource TRANSPORT_TRACE  
0          NaN        Satellite             NaN  
1          NaN        Satellite             NaN  
2          NaN      Coastal AIS             NaN  
3          NaN        Satellite             NaN  
4          NaN        Satellite             NaN

In [13]:
df_sample.timestamp = pd.to_datetime(df_sample.timestamp)

In [14]:
df_sample = df_sample[~df_sample.vesselNextport.isna()]

In [15]:
df_sample.shape

(71721, 13)

In [21]:
col_selected = ["loadingOrder","timestamp","carrierName","longitude",
                "latitude","vesselMMSI","speed","direction","vesselNextport"]
df_sample_short = df_sample.sort_values(["loadingOrder","timestamp"])[col_selected]
df_sample_short

loadingOrder                 timestamp carrierName   longitude  \
31170  AE636773257777 2019-01-01 14:53:06+00:00      KJIHXD  114.241933   
31427  AE636773257777 2019-01-01 14:59:18+00:00      KJIHXD  114.220333   
31783  AE636773257777 2019-01-01 15:09:54+00:00      KJIHXD  114.199650   
32068  AE636773257777 2019-01-01 15:17:54+00:00      KJIHXD  114.187717   
32173  AE636773257777 2019-01-01 15:22:54+00:00      KJIHXD  114.181417   
...               ...                       ...         ...         ...   
96508  ZZ391515915222 2019-01-02 20:00:41+00:00      UQCRKD  172.071177   
97344  ZZ391515915222 2019-01-02 20:29:10+00:00      UQCRKD  172.217653   
97570  ZZ391515915222 2019-01-02 20:34:11+00:00      UQCRKD  172.242757   
98927  ZZ391515915222 2019-01-02 21:04:12+00:00      UQCRKD  172.395478   
99153  ZZ391515915222 2019-01-02 21:09:17+00:00      UQCRKD  172.421457   

        latitude   vesselMMSI  speed  direction vesselNextport  
31170  22.143550  O8605873484     27      30530      HONG KONG  
31427  22.158267  O8605873484     24      31210      HONG KONG  
31783  22.183000  O8605873484     15      32360      HONG KONG  
32068  22.197833  O8605873484     15      32330      HONG KONG  
32173  22.205883  O8605873484     12      32380      HONG KONG  
...          ...          ...    ...        ...            ...  
96508 -33.788287  H9283003522     30      11100       AUCKLAND  
97344 -33.834693  H9283003522     30      11550       AUCKLAND  
97570 -33.844177  H9283003522     30      11470       AUCKLAND  
98927 -33.898222  H9283003522     30      11130       AUCKLAND  
99153 -33.907312  H9283003522     30      11080       AUCKLAND  

[71721 rows x 9 columns]

In [22]:
df_sample_short["time_diff"]=df_sample_short["timestamp"].diff()
df_sample_short["long_diff"]=df_sample_short["longitude"].diff()
df_sample_short["lat_diff"]=df_sample_short["latitude"].diff()
col_selected = ["loadingOrder","time_diff","long_diff",
                "lat_diff","speed","direction"]
df_sample_short = df_sample_short[col_selected]

In [23]:
df_sample_short["dist_move"]=np.sqrt(df_sample_short["long_diff"]**2+df_sample_short["lat_diff"]**2)
df_sample_short.head(30)

F:\Program\Anaconda\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


loadingOrder         time_diff  long_diff   lat_diff  speed  \
31170  AE636773257777               NaT        NaN        NaN     27   
31427  AE636773257777          00:06:12  -0.021600   0.014717     24   
31783  AE636773257777          00:10:36  -0.020683   0.024733     15   
32068  AE636773257777          00:08:00  -0.011933   0.014833     15   
32173  AE636773257777          00:05:00  -0.006300   0.008050     12   
32326  AE636773257777          00:04:01  -0.006050   0.005034     13   
32478  AE636773257777          00:06:00  -0.010450   0.007716     14   
32571  AE636773257777          00:03:58  -0.007067   0.005384     14   
32672  AE636773257777          00:04:00  -0.007267   0.005583     15   
32807  AE636773257777          00:04:02  -0.008400   0.006267     18   
32939  AE636773257777          00:05:15  -0.009950   0.006616     22   
33031  AE636773257777          00:04:19  -0.011033   0.010850     22   
33116  AE636773257777          00:03:20  -0.005217   0.009717     20   
33246  AE636773257777          00:04:31  -0.006233   0.011550     17   
33382  AE636773257777          00:03:21  -0.003683   0.007633     16   
33513  AE636773257777          00:04:40  -0.004850   0.009867     15   
33602  AE636773257777          00:03:28  -0.003900   0.007200     16   
33776  AE636773257777          00:04:39   0.000116   0.009983     14   
33958  AE636773257777          00:03:31   0.004834   0.006817     16   
34397  AE636773257777          00:03:39   0.007850   0.004917     14   
34741  AE636773257777          00:04:32   0.007633   0.004783     10   
34865  AE636773257777          00:03:38   0.002317   0.004100      6   
35001  AE636773257777          00:03:59  -0.000634   0.002600      4   
35233  AE636773257777          00:04:33  -0.001800   0.002150      3   
35564  AE636773257777          00:03:28  -0.000716   0.001483      2   
35731  AE636773257777          00:04:31  -0.000384   0.000284      0   
166    AF630883821362 -1 days +07:14:38  -7.205433 -28.281515      2   
1441   AF630883821362          00:41:40   0.004943   0.012522      1   
1878   AF630883821362          00:14:56   0.001344   0.004256      2   
4734   AF630883821362          01:31:21   0.014513   0.025845      2   

       direction  dist_move  
31170      30530        NaN  
31427      31210   0.026137  
31783      32360   0.032241  
32068      32330   0.019037  
32173      32380   0.010222  
32326      30720   0.007870  
32478      30890   0.012990  
32571      30920   0.008884  
32672      30940   0.009164  
32807      30670   0.010480  
32939      30640   0.011949  
33031      33320   0.015474  
33116      33350   0.011029  
33246      33580   0.013125  
33382      33560   0.008475  
33513      33470   0.010995  
33602      33000   0.008188  
33776       2150   0.009984  
33958       4700   0.008357  
34397       5790   0.009263  
34741       4280   0.009008  
34865       1230   0.004709  
35001      32710   0.002676  
35233      33050   0.002804  
35564      33580   0.001647  
35731      27460   0.000478  
166          500  29.184968  
1441       35750   0.013462  
1878         310   0.004463  
4734        2700   0.029641

In [24]:
df_sample_short[70:90]

loadingOrder         time_diff  long_diff   lat_diff  speed  \
22508  AF630883821362          00:12:00  -0.000183  -0.000152      0   
22699  AF630883821362          00:06:04   0.000023  -0.000002      0   
22860  AF630883821362          00:05:56  -0.000057   0.000127      0   
23003  AF630883821362          00:06:01   0.000294  -0.000027      0   
23163  AF630883821362          00:06:00  -0.000109  -0.000008      0   
23598  AF630883821362          00:12:00  -0.000003  -0.000147      0   
23925  AF630883821362          00:06:00  -0.000253   0.000069      0   
24314  AF630883821362          00:06:00   0.000130   0.000036      0   
24613  AF630883821362          00:06:00   0.000176   0.000047      0   
24787  AF630883821362          00:05:59  -0.000066  -0.000067      0   
25315  AF630883821362          00:12:00   0.000010   0.000005      0   
25544  AF630883821362          00:03:01  -0.000089   0.000004      0   
25969  AF630883821362          00:05:59   0.000034  -0.000022      0   
4754   AF641280380594 -1 days +13:57:31 -71.990569  32.789617     32   
5895   AF641280380594          00:26:17  -0.080773   0.105413     32   
6190   AF641280380594          00:10:05  -0.030267   0.041120     32   
17492  AF641280380594          05:50:17  -1.144646   1.084538     25   
18258  AF641280380594          00:29:00  -0.087754   0.071515     24   
18450  AF641280380594          00:05:45  -0.016800   0.013894     24   
23125  AF641280380594          02:03:21  -0.309760   0.349493     25   

       direction  dist_move  
22508      25150   0.000238  
22699      25150   0.000023  
22860      25150   0.000139  
23003      25150   0.000295  
23163      25150   0.000109  
23598      23300   0.000147  
23925      23300   0.000262  
24314      23300   0.000135  
24613      23300   0.000182  
24787      23300   0.000094  
25315      23300   0.000011  
25544      23300   0.000089  
25969      23300   0.000040  
4754       32510  79.106264  
5895       32580   0.132801  
6190       32640   0.051058  
17492      30410   1.576844  
18258      31360   0.113204  
18450      31330   0.021801  
23125      33000   0.467008

TRANSPORT_TRACE      172 non-null object: 1.72% non-null values. 不值得使用这个数据

In [25]:
df_sample.timestamp = pd.to_datetime(df_sample.timestamp)

In [26]:
df_grouped = df_sample.groupby("loadingOrder").apply(lambda x: x.sort_values('timestamp'))

In [27]:
df_grouped.loc[df_grouped.index.levels[0][3]]

loadingOrder carrierName                 timestamp   longitude  \
169    AF801994078018      JONOCD 2019-01-01 00:04:57+00:00  106.911100   
1443   AF801994078018      JONOCD 2019-01-01 00:46:37+00:00  106.916043   
1880   AF801994078018      JONOCD 2019-01-01 01:01:33+00:00  106.917387   
4736   AF801994078018      JONOCD 2019-01-01 02:32:54+00:00  106.931900   
5769   AF801994078018      JONOCD 2019-01-01 02:55:45+00:00  106.922782   
5859   AF801994078018      JONOCD 2019-01-01 02:59:04+00:00  106.919493   
7425   AF801994078018      JONOCD 2019-01-01 03:46:10+00:00  106.914695   
7673   AF801994078018      JONOCD 2019-01-01 03:56:25+00:00  106.907722   
8457   AF801994078018      JONOCD 2019-01-01 04:25:01+00:00  106.886000   
8947   AF801994078018      JONOCD 2019-01-01 04:42:31+00:00  106.882892   
9066   AF801994078018      JONOCD 2019-01-01 04:45:50+00:00  106.883280   
9112   AF801994078018      JONOCD 2019-01-01 04:47:46+00:00  106.883092   
9479   AF801994078018      JONOCD 2019-01-01 04:55:25+00:00  106.879797   
9686   AF801994078018      JONOCD 2019-01-01 05:00:31+00:00  106.881613   
9898   AF801994078018      JONOCD 2019-01-01 05:06:05+00:00  106.882665   
10124  AF801994078018      JONOCD 2019-01-01 05:11:45+00:00  106.882662   
10493  AF801994078018      JONOCD 2019-01-01 05:18:45+00:00  106.882500   
10634  AF801994078018      JONOCD 2019-01-01 05:21:04+00:00  106.882378   
10747  AF801994078018      JONOCD 2019-01-01 05:25:05+00:00  106.883012   
10912  AF801994078018      JONOCD 2019-01-01 05:33:25+00:00  106.889520   
11137  AF801994078018      JONOCD 2019-01-01 05:39:51+00:00  106.893728   
11303  AF801994078018      JONOCD 2019-01-01 05:43:26+00:00  106.894855   
11540  AF801994078018      JONOCD 2019-01-01 05:53:05+00:00  106.895763   
16650  AF801994078018      JONOCD 2019-01-01 08:32:36+00:00  106.896133   
16866  AF801994078018      JONOCD 2019-01-01 08:38:35+00:00  106.895963   
17042  AF801994078018      JONOCD 2019-01-01 08:44:36+00:00  106.895957   
17248  AF801994078018      JONOCD 2019-01-01 08:50:36+00:00  106.895863   
17388  AF801994078018      JONOCD 2019-01-01 08:56:35+00:00  106.896007   
17582  AF801994078018      JONOCD 2019-01-01 09:02:35+00:00  106.896060   
17784  AF801994078018      JONOCD 2019-01-01 09:08:36+00:00  106.896137   
17968  AF801994078018      JONOCD 2019-01-01 09:14:35+00:00  106.896213   
18109  AF801994078018      JONOCD 2019-01-01 09:20:36+00:00  106.896083   
18375  AF801994078018      JONOCD 2019-01-01 09:32:36+00:00  106.896120   
18913  AF801994078018      JONOCD 2019-01-01 09:44:37+00:00  106.896032   
19052  AF801994078018      JONOCD 2019-01-01 09:50:36+00:00  106.896025   
19224  AF801994078018      JONOCD 2019-01-01 09:56:36+00:00  106.896120   
19309  AF801994078018      JONOCD 2019-01-01 10:02:35+00:00  106.896108   
19468  AF801994078018      JONOCD 2019-01-01 10:08:35+00:00  106.896035   
20042  AF801994078018      JONOCD 2019-01-01 10:17:36+00:00  106.896072   
20711  AF801994078018      JONOCD 2019-01-01 10:26:35+00:00  106.896040   
21189  AF801994078018      JONOCD 2019-01-01 10:38:37+00:00  106.895975   
21378  AF801994078018      JONOCD 2019-01-01 10:44:36+00:00  106.895982   
21782  AF801994078018      JONOCD 2019-01-01 10:53:36+00:00  106.895982   
22078  AF801994078018      JONOCD 2019-01-01 11:02:35+00:00  106.896155   
22511  AF801994078018      JONOCD 2019-01-01 11:14:35+00:00  106.895972   
22702  AF801994078018      JONOCD 2019-01-01 11:20:39+00:00  106.895995   
22863  AF801994078018      JONOCD 2019-01-01 11:26:35+00:00  106.895938   
23006  AF801994078018      JONOCD 2019-01-01 11:32:36+00:00  106.896232   
23166  AF801994078018      JONOCD 2019-01-01 11:38:36+00:00  106.896123   
23601  AF801994078018      JONOCD 2019-01-01 11:50:36+00:00  106.896120   
23921  AF801994078018      JONOCD 2019-01-01 11:56:36+00:00  106.895867   
24317  AF801994078018      JONOCD 2019-01-01 12:02:36+00:00  106.895997   
24609  AF801994078018      JONOCD 

vesselNextportETA 这个字段可以丢弃，用处不大。

据反映，会出现换船的情况。

In [28]:
ship_number = df_sample.groupby('loadingOrder')["vesselMMSI"].nunique()

In [29]:
ship_number.max()

1

In [30]:
transfer_times = df_sample.groupby('loadingOrder')["vesselNextport"].nunique()

In [31]:
transfer_times[transfer_times==2]

loadingOrder
AE636773257777    2
AF630883821362    2
AF641280380594    2
AF801994078018    2
AT211137428479    2
                 ..
YP986032223572    2
YT265857616487    2
YU362992712613    2
YY327315496040    2
YZ946549358503    2
Name: vesselNextport, Length: 110, dtype: int64

## 修正port表格

修正思路：
用transport_node_id作为唯一标识物，来确定某个货物到达的位置

问题：
- [X] 存在NODE_ID缺失的情况。被证明是数据歪了，已用ID的前面两列进行填充
- [X] 相同ID对应的经纬度信息不一样。已经手动修改了。
- [ ] 同样的TRANS_NODE_NAME对应多个ID，在不同的位置。怎么办？比如OTP，这个名字对应了四条记录，其中有三条的位置非常接近，有一条位置很奇怪。

In [33]:
df_port = pd.read_csv("data/port.csv")
df_port

TRANS_NODE_NAME   LONGITUDE   LATITUDE      COUNTRY  \
0           NP1_POK    83.975602  28.204030        Nepal   
1           NP1_NEP    81.361656  28.023097        India   
2           NP1_NWP    84.399719  27.696448        Nepal   
3          NP1_KMDU    85.230904  27.410841        Nepal   
4           NP1_EDR    87.270927  26.587093        Nepal   
...              ...         ...        ...          ...   
2451            Maza   67.112458  36.735551  Afghanistan   
2452           LKCOL   79.861243   6.927079    Sri Lanka   
2453           MYPKN  101.392830   2.999852     Malaysia   
2454           CRPLI  -83.041508   9.991311   Costa Rica   
2455    HAKATA_JAPAN  130.441447  33.590342        Japan   

                           STATE            CITY      REGION  \
0     Western Development Region         Pokhara         NaN   
1                  Uttar Pradesh   Parwani Gauri         NaN   
2     Western Development Region       Gaindakot         NaN   
3     Central Development Region         Thingan         NaN   
4     Eastern Development Region             NaN         NaN   
...                          ...             ...         ...   
2451                       Balkh  Mazar-e Sharif         NaN   
2452            Western Province         Colombo  Colombo 10   
2453                    Selangor           Klang  Kawasan 13   
2454                        Lim?            Lim?         NaN   
2455                     Fukuoka         Fukuoka   Hakata-ku   

                                                ADDRESS  PORT_CODE  \
0          Janapriya Marga Lane 2, Pokhara 33700, Nepal        NaN   
1     Nanpara Rd, Parwani Gauri, Uttar Pradesh 27185...        NaN   
2           Ganesh Mandir Marga, Gaindakot 33000, Nepal        NaN   
3                    Unnamed Road, Thingan 44100, Nepal        NaN   
4                      Madhelee Duhabi Rd, 56709, Nepal        NaN   
...                                                 ...        ...   
2451          Unnamed Road, Mazar-e Sharif, Afghanistan        NaN   
2452              Gamini Hall, Colombo 01000, Sri Lanka        NaN   
2453  1, Persiaran Raja Muda Musa, Kawasan 13, 41100...        NaN   
2454        Avenida Macarela, El Bamb? Lim?, Costa Rica        NaN   
2455                                                NaN        NaN   

      TRANSPORT_NODE_ID  
0              117000.0  
1              118000.0  
2              119000.0  
3              120000.0  
4              121000.0  
...                 ...  
2451        281615000.0  
2452        282214000.0  
2453        282215000.0  
2454        283015000.0  
2455        284217000.0  

[2456 rows x 10 columns]

In [34]:
df_port.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2456 entries, 0 to 2455
Data columns (total 10 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   TRANS_NODE_NAME    2456 non-null   object 
 1   LONGITUDE          2456 non-null   float64
 2   LATITUDE           2456 non-null   float64
 3   COUNTRY            2456 non-null   object 
 4   STATE              1778 non-null   object 
 5   CITY               2163 non-null   object 
 6   REGION             654 non-null    object 
 7   ADDRESS            1942 non-null   object 
 8   PORT_CODE          25 non-null     float64
 9   TRANSPORT_NODE_ID  2420 non-null   float64
dtypes: float64(4), object(6)
memory usage: 192.0+ KB


假定港口ID是唯一的，但是这里有的港口没有ID，什么情况？？？

In [35]:
df_port[df_port.TRANSPORT_NODE_ID.isna()]

TRANS_NODE_NAME  LONGITUDE   LATITUDE                   COUNTRY  \
23          PA1_DHL  -79.408150   9.063167                    Panama   
69           SANJOSE -84.115105   9.944842                Costa Rica   
70               SJO -84.115105   9.944842                Costa Rica   
73               ICE -84.100393   9.945933                Costa Rica   
182            BRSSZ -46.284020 -23.954513                    Brazil   
228              BGF  18.519592   4.399705  Central African Republic   
229            ARRAZ -70.138945 -20.211789                     Chile   
230            CLIQQ -70.138945 -20.211789                     Chile   
278           BOGOTA -74.143677   4.700564                  Colombia   
279              BOG -74.143677   4.700564                  Colombia   
282              COL -74.144668   4.701156                  Colombia   
587            PABLB -79.568040   8.966497                    Panama   
596              PTY -79.386687   9.067888                    Panama   
645            STTMS   6.736317   0.347019     Sao Tome and Principe   
646              TMS   6.712369   0.377950     Sao Tome and Principe   
987       CO_SCHENKE -74.167191   4.676915                  Colombia   
1059          532926 -79.419136   9.063957                    Panama   
1152      SU_SCHENKE -74.165945   4.677109                  Colombia   
1159      SU1_OFFICE -74.055187   4.695402                  Colombia   
1172          PA_TLF -79.474959   9.010959                    Panama   
1198      PR_CROSSDO -49.195009 -25.412519                    Brazil   
1200       RDA_PA_XD -48.505445  -1.460275                    Brazil   
1202      CE_CROSSDO -38.503879  -3.828333                    Brazil   
1205        CLARO_PR -49.190084 -25.529713                    Brazil   
1221       RDA_CE_XD -38.508227  -3.808879                    Brazil   
1228             LSC -79.875079   9.341103                    Panama   
1273           PAPA2 -79.588265   8.984147                    Panama   
1291           BRPNG -48.507342 -25.514172                    Brazil   
1327      PAN-STOP 1 -79.812729   8.853239                    Panama   
1415           PAPTY -79.552002   8.966641                    Panama   
2054        FR  TEST   4.610773  46.910182                    France   
2189           TGFLW   1.278569   6.137216                      Togo   
2210             MLE  73.530701   4.189310                  Maldives   
2295           PABAL -79.566958   8.961348                    Panama   
2296           PACOL -79.888336   9.356716                    Panama   
2319           PAPAN -79.535405   8.958647                    Panama   

                            STATE  \
23                   Panam?Panam?   
69               San Jos?San Jos?   
70               San Jos?San Jos?   
73               San Jos?San Jos?   
182                      S? Paulo   
228          K??gb??t?Bang?Bangui   
229      Regi? de Tarapac?Iquique   
230      Regi? de Tarapac?Iquique   
278                  Bogot?Bogota   
279                  Bogot?Bogota   
282            Bogot?Isle Of Coll   
587                  Panam?Balboa   
596             Panam?Panama City   
645               S? Tom?Sao Tome   
646            S? Tom?Sao Tome Is   
987           Bogot?Bogot?Fontib?   
1059                 Panam?Panam?   
1152          Bogot?Bogot?Fontib?   
1159           Bogot?Bogot?Usaqu?   
1172                 Panam?Panam?   
1198                Paran?Pinhais   
1200                     Par?Bel?   
1202               Cear?Fortaleza   
1205     Paran?S? Jos?dos Pinhais   
1221               Cear?Fortaleza   
1228            Provincia de Col?   
1273                 Panam?Cocoli   
1291              Paran?Paranagua   
1327            Panam?La Chorrera   
1415    Panam?Panama City           
2054  Bourgogne-Franche-Comt?Cr?t   
2189               R?ion maritime   
2210                          NaN   
2295                 Panam?Panam?   
2296            Provincia de Col?   
2319    

查看数据，是因为有的ID位置歪了，到了前面。会不会是编码问题？

In [36]:
import chardet

with open("data/port.csv", 'rb') as fin:
    encoding_type = chardet.detect(fin.read(70))['encoding']
encoding_type

'ascii'

In [38]:
df_port = pd.read_csv("data/port.csv",encoding="ascii")
df_port.head()

TRANS_NODE_NAME  LONGITUDE   LATITUDE COUNTRY                       STATE  \
0        NP1_POK   83.975602  28.204030   Nepal  Western Development Region   
1        NP1_NEP   81.361656  28.023097   India               Uttar Pradesh   
2        NP1_NWP   84.399719  27.696448   Nepal  Western Development Region   
3       NP1_KMDU   85.230904  27.410841   Nepal  Central Development Region   
4        NP1_EDR   87.270927  26.587093   Nepal  Eastern Development Region   

            CITY REGION                                            ADDRESS  \
0        Pokhara    NaN       Janapriya Marga Lane 2, Pokhara 33700, Nepal   
1  Parwani Gauri    NaN  Nanpara Rd, Parwani Gauri, Uttar Pradesh 27185...   
2      Gaindakot    NaN        Ganesh Mandir Marga, Gaindakot 33000, Nepal   
3        Thingan    NaN                 Unnamed Road, Thingan 44100, Nepal   
4            NaN    NaN                   Madhelee Duhabi Rd, 56709, Nepal   

   PORT_CODE  TRANSPORT_NODE_ID  
0        NaN           117000.0  
1        NaN           118000.0  
2        NaN           119000.0  
3        NaN           120000.0  
4        NaN           121000.0

In [39]:
df_port.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2456 entries, 0 to 2455
Data columns (total 10 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   TRANS_NODE_NAME    2456 non-null   object 
 1   LONGITUDE          2456 non-null   float64
 2   LATITUDE           2456 non-null   float64
 3   COUNTRY            2456 non-null   object 
 4   STATE              1778 non-null   object 
 5   CITY               2163 non-null   object 
 6   REGION             654 non-null    object 
 7   ADDRESS            1942 non-null   object 
 8   PORT_CODE          25 non-null     float64
 9   TRANSPORT_NODE_ID  2420 non-null   float64
dtypes: float64(4), object(6)
memory usage: 192.0+ KB


In [40]:
df_port.loc[69:74]

TRANS_NODE_NAME  LONGITUDE   LATITUDE               COUNTRY  \
69         SANJOSE -84.115105   9.944842            Costa Rica   
70             SJO -84.115105   9.944842            Costa Rica   
71  PADC Warehouse -79.874583   9.340270                Panama   
72   HW Costa Rica -84.113110  10.000337            Costa Rica   
73             ICE -84.100393   9.945933            Costa Rica   
74             DSC  55.139008  24.891886  United Arab Emirates   

                STATE                                               CITY  \
69   San Jos?San Jos?  Paseo de la Segunda Rep?lica, San Jos? Costa Rica   
70   San Jos?San Jos?  Paseo de la Segunda Rep?lica, San Jos? Costa Rica   
71  Provincia de Col?                                               Col?   
72            Heredia                                            Heredia   
73   San Jos?San Jos?                      Calle 48, San Jos? Costa Rica   
74              Dubai                                              Dubai   

   REGION                                      ADDRESS  PORT_CODE  \
69    NaN                                      1683000        NaN   
70    NaN                                      1683000        NaN   
71    NaN              Carr. Trans?tmica, Col?, Panama        NaN   
72    NaN                   Av. 3, Heredia, Costa Rica        NaN   
73    NaN                                      1733000        NaN   
74    NaN  Unnamed Road - Dubai - United Arab Emirates        NaN   

    TRANSPORT_NODE_ID  
69                NaN  
70                NaN  
71          1703000.0  
72          1713000.0  
73                NaN  
74          1753000.0

In [41]:
relocate_id = []
for i in df_port.ADDRESS:
    try:
        node_id = float(i)
        
        if np.isnan(node_id):
            #print(i)
            continue
        relocate_id.append(node_id)
    except:
        continue
len(relocate_id)

11

In [42]:
relocate_id

[187000.0,
 1683000.0,
 1683000.0,
 1733000.0,
 3303000.0,
 3374000.0,
 3473000.0,
 3479000.0,
 3491000.0,
 211826000.0,
 212029000.0]

In [43]:
relocate_id_two = []
for i in df_port.PORT_CODE:
    try:
        node_id = float(i)
        
        if np.isnan(node_id):
            #print(i)
            continue
        relocate_id_two.append(node_id)
    except:
        continue
len(relocate_id_two)

25

In [44]:
relocate_id_two

[2462000.0,
 2511000.0,
 2512000.0,
 2512000.0,
 2573000.0,
 2573000.0,
 2576000.0,
 2895000.0,
 2905000.0,
 2957000.0,
 2958000.0,
 3517000.0,
 3519000.0,
 3521000.0,
 3524000.0,
 3540000.0,
 3546000.0,
 3663000.0,
 3684000.0,
 3763000.0,
 4434000.0,
 170615000.0,
 205051000.0,
 211022000.0,
 211827000.0]

数字是对的，证明id的位置真的是歪了。它是唯一的。

In [47]:
id_port_lines

Int64Index([ 182,  228,  229,  230,  278,  279,  282,  587,  596,  645,  646,
            1198, 1200, 1202, 1205, 1221, 1228, 1273, 1291, 1327, 1415, 2054,
            2189, 2210, 2296],
           dtype='int64')

In [46]:
id_port_lines = df_port.index[~df_port.PORT_CODE.isna()]
df_port["TRANSPORT_NODE_ID"].loc[id_port_lines] = df_port["PORT_CODE"].loc[id_port_lines]
id_wrong_lines = df_port.index[df_port.TRANSPORT_NODE_ID.isna()]
df_port["TRANSPORT_NODE_ID"].loc[id_wrong_lines] = df_port["ADDRESS"].loc[id_wrong_lines]
df_port.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2456 entries, 0 to 2455
Data columns (total 10 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   TRANS_NODE_NAME    2456 non-null   object 
 1   LONGITUDE          2456 non-null   float64
 2   LATITUDE           2456 non-null   float64
 3   COUNTRY            2456 non-null   object 
 4   STATE              1778 non-null   object 
 5   CITY               2163 non-null   object 
 6   REGION             654 non-null    object 
 7   ADDRESS            1942 non-null   object 
 8   PORT_CODE          25 non-null     float64
 9   TRANSPORT_NODE_ID  2456 non-null   object 
dtypes: float64(3), object(7)
memory usage: 192.0+ KB


F:\Program\Anaconda\lib\site-packages\pandas\core\indexing.py:671: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._setitem_with_indexer(indexer, value)


In [48]:
df_port_ex = df_port.groupby("TRANSPORT_NODE_ID")["TRANS_NODE_NAME"].count()
df_port_ex = df_port_ex[df_port_ex>1]
df_port_ex.shape

(56,)

In [49]:
df_port[df_port.TRANSPORT_NODE_ID == df_port_ex.index[14]]

TRANS_NODE_NAME  LONGITUDE  LATITUDE COUNTRY              STATE  \
582     COLON FREE TRADE ZONE -79.863599  9.358666  Panama  Provincia de Col?   
583  COLON CONTAINER TERMINAL -79.863599  9.358666  Panama  Provincia de Col?   
584        COLONFREETRADEZONE -79.863599  9.358666  Panama  Provincia de Col?   
585    COLONCONTAINERTERMINAL -79.863599  9.358666  Panama  Provincia de Col?   
586                       ONX -79.863599  9.358666  Panama  Provincia de Col?   

      CITY REGION                     ADDRESS  PORT_CODE TRANSPORT_NODE_ID  
582  Colon    NaN  Unnamed Road, Col?, Panama        NaN         2.894e+06  
583  Colon    NaN  Unnamed Road, Col?, Panama        NaN         2.894e+06  
584  Colon    NaN  Unnamed Road, Col?, Panama        NaN         2.894e+06  
585  Colon    NaN  Unnamed Road, Col?, Panama        NaN         2.894e+06  
586  Colon    NaN  Unnamed Road, Col?, Panama        NaN         2.894e+06

检查一下，NODE_ID相同的记录，他们的经纬度信息是不是一样的

In [50]:
df_ex = None

def NumUnique(x):
    return len(np.unique(x))
for i in df_port_ex.index:
    df_sub = df_port[df_port.TRANSPORT_NODE_ID == i]
    df_res = df_sub[["LONGITUDE","LATITUDE","COUNTRY"]].agg(NumUnique)
    df_res["ID"]=i
    if df_ex is None:
        df_ex = df_res
    else:
        df_ex = pd.concat([df_ex,df_res],axis=1)
    

In [51]:
df_ex2 = pd.DataFrame(df_ex.values.T, index=df_ex.columns, columns=df_ex.index)
df_ex2 = df_ex2.set_index("ID")
df_ex2.head()

LONGITUDE LATITUDE COUNTRY
ID                                  
1288000.0         1        1       1
2294000.0         1        1       1
2401000.0         1        1       1
2415000.0         1        1       1
2512000.0         1        1       1

In [52]:
df_ex2['row_sum'] = df_ex2.apply(lambda x: x.sum(), axis=1)
df_ex2[df_ex2['row_sum']>3]

LONGITUDE LATITUDE COUNTRY  row_sum
ID                                             
2580000.0           1        1       2      4.0
235014000.0         2        2       1      5.0

In [53]:
df_port[df_port.TRANSPORT_NODE_ID==2580000]

TRANS_NODE_NAME  LONGITUDE  LATITUDE COUNTRY         STATE  \
286  POINTENOIRE_CONGO  11.878405 -4.803998   CONGO  Pointe-Noire   
287                PNR  11.878405 -4.803998   Congo  Pointe-Noire   

             CITY REGION                                            ADDRESS  \
286  Pointe Noire    NaN  Unnamed Road, Pointe-Noire, Republic of the Congo   
287  Pointe Noire    NaN  Unnamed Road, Pointe-Noire, Republic of the Congo   

     PORT_CODE TRANSPORT_NODE_ID  
286        NaN          2.58e+06  
287        NaN          2.58e+06

In [54]:
df_port.loc[287,"COUNTRY"]="CONGO"
df_port.loc[287]

TRANS_NODE_NAME                                                    PNR
LONGITUDE                                                      11.8784
LATITUDE                                                        -4.804
COUNTRY                                                          CONGO
STATE                                                     Pointe-Noire
CITY                                                      Pointe Noire
REGION                                                             NaN
ADDRESS              Unnamed Road, Pointe-Noire, Republic of the Congo
PORT_CODE                                                          NaN
TRANSPORT_NODE_ID                                             2.58e+06
Name: 287, dtype: object

In [55]:
df_port[df_port.TRANSPORT_NODE_ID==235014000]

TRANS_NODE_NAME   LONGITUDE   LATITUDE COUNTRY             STATE  \
2400             VAN -130.320819  54.315037  Canada               NaN   
2401          CA VAN -123.103178  49.312004  Canada  British Columbia   
2402           CAVAN -123.103178  49.312004  Canada  British Columbia   

                 CITY REGION  \
2400              NaN    NaN   
2401  North Vancouver    NaN   
2402  North Vancouver    NaN   

                                                ADDRESS  PORT_CODE  \
2400                                                NaN        NaN   
2401  Unnamed Road, North Vancouver, BC V7P 2R2, Canada        NaN   
2402  Unnamed Road, North Vancouver, BC V7P 2R2, Canada        NaN   

     TRANSPORT_NODE_ID  
2400       2.35014e+08  
2401       2.35014e+08  
2402       2.35014e+08

经过查询[link1](https://baike.baidu.com/item/%E6%B8%A9%E5%93%A5%E5%8D%8E%E6%B8%AF/3174223?fr=aladdin), [link2](https://www.earthol.com/)，我认为2400的数据是错误的。将其求改为与2401和2402一致。

In [56]:
df_port.loc[2400,"LONGITUDE"]=df_port.loc[2401,"LONGITUDE"]
df_port.loc[2400,"LATITUDE"]=df_port.loc[2401,"LATITUDE"]

In [57]:
df_ex = None

def NumUnique(x):
    return len(np.unique(x))
for i in df_port_ex.index:
    df_sub = df_port[df_port.TRANSPORT_NODE_ID == i]
    df_res = df_sub[["LONGITUDE","LATITUDE","COUNTRY"]].agg(NumUnique)
    df_res["ID"]=i
    if df_ex is None:
        df_ex = df_res
    else:
        df_ex = pd.concat([df_ex,df_res],axis=1)
df_ex2 = pd.DataFrame(df_ex.values.T, index=df_ex.columns, columns=df_ex.index)
df_ex2 = df_ex2.set_index("ID")
df_ex2['row_sum'] = df_ex2.apply(lambda x: x.sum(), axis=1)
df_ex2[df_ex2['row_sum']>3]

Empty DataFrame
Columns: [LONGITUDE, LATITUDE, COUNTRY, row_sum]
Index: []

In [58]:
df_port.to_csv("data/port_fixed_0612.csv")

In [59]:
df_port=pd.read_csv("data/port_fixed_0612.csv",index_col=0)
df_port.head()

TRANS_NODE_NAME  LONGITUDE   LATITUDE COUNTRY                       STATE  \
0        NP1_POK   83.975602  28.204030   Nepal  Western Development Region   
1        NP1_NEP   81.361656  28.023097   India               Uttar Pradesh   
2        NP1_NWP   84.399719  27.696448   Nepal  Western Development Region   
3       NP1_KMDU   85.230904  27.410841   Nepal  Central Development Region   
4        NP1_EDR   87.270927  26.587093   Nepal  Eastern Development Region   

            CITY REGION                                            ADDRESS  \
0        Pokhara    NaN       Janapriya Marga Lane 2, Pokhara 33700, Nepal   
1  Parwani Gauri    NaN  Nanpara Rd, Parwani Gauri, Uttar Pradesh 27185...   
2      Gaindakot    NaN        Ganesh Mandir Marga, Gaindakot 33000, Nepal   
3        Thingan    NaN                 Unnamed Road, Thingan 44100, Nepal   
4            NaN    NaN                   Madhelee Duhabi Rd, 56709, Nepal   

   PORT_CODE  TRANSPORT_NODE_ID  
0        NaN           117000.0  
1        NaN           118000.0  
2        NaN           119000.0  
3        NaN           120000.0  
4        NaN           121000.0

In [60]:
df_port[df_port.TRANS_NODE_NAME=="MYPKG"]

TRANS_NODE_NAME   LONGITUDE  LATITUDE   COUNTRY     STATE  \
42             MYPKG  101.361204  3.034709  Malaysia  Selangor   
2438           MYPKG  101.392830  2.999852  Malaysia  Selangor   

                 CITY           REGION  \
42    Pelabuhan Klang  Pelabuhan Utara   
2438            Klang       Kawasan 13   

                                                ADDRESS  PORT_CODE  \
42    Jalan Kerisi, Pelabuhan Utara, 42000 Pelabuhan...        NaN   
2438  1, Persiaran Raja Muda Musa, Kawasan 13, 41100...        NaN   

      TRANSPORT_NODE_ID  
42             636000.0  
2438        272214000.0

In [61]:
df_name_cnt = df_port.groupby("TRANS_NODE_NAME")["TRANSPORT_NODE_ID"].count()
df_name_cnt[df_name_cnt>1]

TRANS_NODE_NAME
/                       4
BUH                     2
CNHKG                   2
DSC                    28
ESALG                   2
ESC                     3
Guatemala warehouse     2
HAV                     2
HRRIJ                   3
HTM.??                  2
HTM.????                4
IHW                     2
IST                     2
JAKARTA                 2
LMS                     2
LSC                     4
MIT                     2
MYBKI                   2
MYPKG                   2
OTP                     4
PIRAEUS_GREECE          2
Polish border           2
RIX                     2
SANTA CRUZ_BOLIVIA      2
SJO                     2
Smonlensk CC point      2
South  WH of MGF        2
Sulaibiya               2
YUL                     2
YYC                     2
YYZ                     2
ZMLUN                   2
Name: TRANSPORT_NODE_ID, dtype: int64

In [62]:
df_port[df_port.TRANS_NODE_NAME=="OTP"]

TRANS_NODE_NAME  LONGITUDE   LATITUDE  COUNTRY          STATE     CITY  \
1619             OTP  26.083511  44.567187  Romania  Jude?ul Ilfov  Otopeni   
1805             OTP   0.831116  -0.386592      All            NaN      NaN   
2017             OTP  26.084412  44.570731  Romania  Jude?ul Ilfov  Otopeni   
2234             OTP  26.083266  44.574662  Romania  Jude?ul Ilfov  Otopeni   

     REGION                                            ADDRESS  PORT_CODE  \
1619    NaN  ???? Otopeni, Calea Bucure?tilor, ????????????...        NaN   
1805    NaN                                                NaN        NaN   
2017    NaN  nr 291B, Aeroportul Henri Coanda, Calea Bucure...        NaN   
2234    NaN            Cargo Terminal, Otopeni 075100, Romania        NaN   

      TRANSPORT_NODE_ID  
1619         28423000.0  
1805         84814000.0  
2017        129421000.0  
2234        211224000.0

In [63]:
df_port[df_port.TRANS_NODE_NAME=="YYC"]

TRANS_NODE_NAME   LONGITUDE   LATITUDE COUNTRY    STATE     CITY REGION  \
233              YYC -114.007276  51.126461  Canada  Alberta  Calgary    NaN   
2023             YYC -114.007616  51.121503  Canada  Alberta  Calgary    NaN   

                                                ADDRESS  PORT_CODE  \
233              8001 21 St NE, Calgary, AB T2E, Canada        NaN   
2023  Calgary International Airport (YYC), 2000 Airp...        NaN   

      TRANSPORT_NODE_ID  
233           2515000.0  
2023        135815000.0

初步清洗完毕，以后可以直接读取清洗后的数据，并用ID作为唯一地理位置识别代码。

例外：TRANS_NODE_NAME一样，但是ID不一样

- [ ] 下一步：看看经纬度坐标是否何理

## Event数据清洗

清洗建议：
- 去掉EVENT_LOCATION_ID为空的记录
- 去掉非重点事件的记录
- 去掉重复的数据，比如16/17两行完全一样
- 去掉EVENT_LOCATION_ID不在port表格中的列
- 梳理记录之间的正确性

In [64]:
df_event = pd.read_csv("data/loadingOrderEvent.csv")
df_event.head()

loadingOrder              EVENT_CODE EVENT_LOCATION_ID  \
0  QA823082754507   SHIPMENT ONBOARD DATE           YANTIAN   
1  CW715391887729  ESTIMATED ONBOARD DATE            SHEKOU   
2  DY644423792756   DAILY TRACK AND TRACE             CNSHK   
3  DY644423792756               PICKED UP    HONGKONG_CHINA   
4  LT180859796067         ARRIVAL AT PORT             NLRTM   

  EVENT_CONVOLUTION_DATE  
0         2019-1-1 17:00  
1         2019-1-1 17:00  
2         2019-1-1 18:00  
3         2019-1-2 10:00  
4         2019-1-3 14:21

In [65]:
df_event.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 158341 entries, 0 to 158340
Data columns (total 4 columns):
 #   Column                  Non-Null Count   Dtype 
---  ------                  --------------   ----- 
 0   loadingOrder            158341 non-null  object
 1   EVENT_CODE              158341 non-null  object
 2   EVENT_LOCATION_ID       152834 non-null  object
 3   EVENT_CONVOLUTION_DATE  158341 non-null  object
dtypes: object(4)
memory usage: 4.8+ MB


In [66]:
df_event.EVENT_CONVOLUTION_DATE = pd.to_datetime(df_event.EVENT_CONVOLUTION_DATE)
df_event.EVENT_CODE.unique()


array(['SHIPMENT ONBOARD DATE', 'ESTIMATED ONBOARD DATE',
       'DAILY TRACK AND TRACE', 'PICKED UP', 'ARRIVAL AT PORT',
       'CARGO ARRIVAL AT DESTINATION', 'UPDATE SHIPMENT ETA',
       'TRANSIT PORT ATA', 'ESTIMATED ARRIVAL TIME TO PORT',
       'PLANNED PICK UP DATE', 'TRANSIT PORT ETA', 'ARRIVAL AT CFS OR CY',
       'TRANSIT PORT ATD', 'TRANSIT PORT ETD',
       'IMP CUSTOMS CLEARANCE START', 'IMP CUSTOMS CLEARANCE FINISHED',
       'DISCHARGED', 'CONTAINER LOADED ON BOARD', 'RDC ATD',
       'TRANSIT PORT DECLARATION BEGIN', 'TRANSIT PORT CUSTOMS RELEASE',
       'ESTIMATED ARRIVAL TO PORT', 'SHIPMENT MIT INBOUND DATE'],
      dtype=object)

- [X] 弄清楚各个字段都是什么意思
    - SHIPMENT ONBOARD DATE: 实际离开起运港
    - ESTIMATED ONBOARD DATE：预计离开起运港日期
    - DAILY TRACK AND TRACE：日常检测（可省略）
    - TRANSIT PORT ETA：预计到达中转港时间
    - TRANSIT PORT ETD：预计离开中转港时间
    - TRANSIT PORT ATD：实际离开中转港时间
    - PICKED UP：货物被提
    - CARGO ARRIVAL AT DESTINATION：货物到达终点
    - ARRIVAL AT PORT：*实际到达目的港*（非常重要）
    - TRANSIT PORT ATA：实际到达中转港时间
    - PLANNED PICK UP DATE：计划提货时间（可忽略）
    - UPDATE SHIPMENT ETA：更新的预计货运到达时间（可忽略）
    - ESTIMATED ARRIVAL TIME TO PORT：预计到达港口时间（可忽略）
    - ARRIVAL AT CFS OR CY：到达集装箱货运站或集装箱堆场
    - IMP CUSTOMS CLEARANCE START：进口报关开始
    - IMP CUSTOMS CLEARANCE FINISHED：进口报关结束
    - DISCHARGED：卸货时间（可忽略）
    - RDC ATD：区域配送中心实际离开时间（可忽略）
- [X] 保留有效的字段
    - TRANSIT PORT ATD实际离开中转港
    - SHIPMENT ONBOARD DATE实际离开起运港
    - TRANSIT PORT ATA实际到达中转港
    - ARRIVAL AT PORT实际到达目的港
    - (CARGO ARRIVAL AT DESTINATION：货物到达终点)：这个时间通常与ARRIVAL AT PORT很接近

In [67]:
events = df_event.EVENT_CODE.unique()

sample_order = df_event[df_event.EVENT_CODE==events[0]].head(1)["loadingOrder"]
df_event[df_event.loadingOrder==sample_order.values[0]].sort_values("EVENT_CONVOLUTION_DATE")

loadingOrder              EVENT_CODE EVENT_LOCATION_ID  \
0      QA823082754507   SHIPMENT ONBOARD DATE           YANTIAN   
7114   QA823082754507  ESTIMATED ONBOARD DATE            SHEKOU   
11794  QA823082754507   DAILY TRACK AND TRACE    SHENZHEN_CHINA   
8085   QA823082754507        TRANSIT PORT ETA             SAJED   
11795  QA823082754507   DAILY TRACK AND TRACE    SHENZHEN_CHINA   
6041   QA823082754507        TRANSIT PORT ETD             PAMAN   

      EVENT_CONVOLUTION_DATE  
0        2019-01-01 17:00:00  
7114     2019-02-04 09:00:00  
11794    2019-02-18 12:00:00  
8085     2019-02-26 23:00:00  
11795    2019-02-27 11:56:00  
6041     2019-03-07 06:55:00

In [68]:
sample_order = df_event[df_event.EVENT_CODE==events[3]].head(1)["loadingOrder"]
df_event[df_event.loadingOrder==sample_order.values[0]].sort_values("EVENT_CONVOLUTION_DATE")

loadingOrder                    EVENT_CODE          EVENT_LOCATION_ID  \
2     DY644423792756         DAILY TRACK AND TRACE                      CNSHK   
3     DY644423792756                     PICKED UP             HONGKONG_CHINA   
183   DY644423792756                     PICKED UP             SHENZHEN_CHINA   
450   DY644423792756  CARGO ARRIVAL AT DESTINATION             VALENCIA_SPAIN   
449   DY644423792756               ARRIVAL AT PORT                      ESVLC   
734   DY644423792756  CARGO ARRIVAL AT DESTINATION      BUENAVENTURA_COLOMBIA   
436   DY644423792756         DAILY TRACK AND TRACE                        NaN   
1660  DY644423792756        ESTIMATED ONBOARD DATE                      CNSHK   
2334  DY644423792756  CARGO ARRIVAL AT DESTINATION  FELIXSTOWE_UNITED KINGDOM   
3893  DY644423792756                     PICKED UP             SHENZHEN_CHINA   
290   DY644423792756         DAILY TRACK AND TRACE                        NaN   
1569  DY644423792756         DAILY TRACK AND TRACE                        NaN   

     EVENT_CONVOLUTION_DATE  
2       2019-01-01 18:00:00  
3       2019-01-02 10:00:00  
183     2019-01-03 10:30:00  
450     2019-01-05 10:54:00  
449     2019-01-05 11:54:00  
734     2019-01-05 23:00:00  
436     2019-01-08 11:30:00  
1660    2019-01-12 19:00:00  
2334    2019-01-13 03:39:00  
3893    2019-01-17 23:16:00  
290     2019-02-02 17:00:00  
1569    2019-02-11 17:00:00

- [ ] 为什么货物在到达之后，还有那么多记录？？
- [ ] 货物到达目的地CARGO ARRIVAL AT DESTINATION和到达港口ARRIVAL AT PORT的时间之间为什么会有差距？
- [ ] 有的EVENT_LOCATION_ID没有出现在port的表格里（TRANS_NODE_NAME列）。例如：19行：CALLAO，119行JEBEL ALI
- [ ] 为什么一个货物的同一个时间会有多条记录，而且时间不一样？比如126，127，128行

In [69]:
sample_order = df_event[df_event.EVENT_CODE==events[7]].head(1)["loadingOrder"]
df_event[df_event.loadingOrder==sample_order.values[0]].sort_values("EVENT_CONVOLUTION_DATE")

loadingOrder                      EVENT_CODE      EVENT_LOCATION_ID  \
15    MT154287661848                       PICKED UP         SHENZHEN_CHINA   
20    MT154287661848                       PICKED UP      ASUNCION_PARAGUAY   
123   MT154287661848           SHIPMENT ONBOARD DATE                 SHEKOU   
16    MT154287661848                TRANSIT PORT ATA                  ZADUR   
17    MT154287661848                TRANSIT PORT ATA                  ZADUR   
121   MT154287661848    CARGO ARRIVAL AT DESTINATION    JAKARTA SEAPORT_IDN   
122   MT154287661848                 ARRIVAL AT PORT                JAKARTA   
18    MT154287661848           SHIPMENT ONBOARD DATE                  CNHKG   
19    MT154287661848                 ARRIVAL AT PORT                 CALLAO   
119   MT154287661848           SHIPMENT ONBOARD DATE              JEBEL ALI   
933   MT154287661848    CARGO ARRIVAL AT DESTINATION  ROTTERDAM_NETHERLANDS   
2214  MT154287661848                       PICKED UP         SHENZHEN_CHINA   
120   MT154287661848  ESTIMATED ARRIVAL TIME TO PORT              JEBEL ALI   

     EVENT_CONVOLUTION_DATE  
15      2018-12-11 14:07:00  
20      2018-12-28 21:00:00  
123     2018-12-31 17:00:00  
16      2019-01-01 00:30:00  
17      2019-01-01 00:30:00  
121     2019-01-01 11:00:00  
122     2019-01-01 17:00:00  
18      2019-01-02 07:48:00  
19      2019-01-02 17:00:00  
119     2019-01-02 17:00:00  
933     2019-01-07 19:00:00  
2214    2019-01-11 11:09:00  
120     2019-01-19 17:00:00

In [70]:
sample_order = df_event[df_event.EVENT_CODE==events[8]].head(1)["loadingOrder"]
df_event[df_event.loadingOrder==sample_order.values[0]].sort_values("EVENT_CONVOLUTION_DATE")

loadingOrder                      EVENT_CODE      EVENT_LOCATION_ID  \
127    WG589003705305           SHIPMENT ONBOARD DATE                 SHEKOU   
128    WG589003705305           SHIPMENT ONBOARD DATE                 SHEKOU   
126    WG589003705305           SHIPMENT ONBOARD DATE                 SHEKOU   
23     WG589003705305                TRANSIT PORT ATA                 PIRAEU   
22     WG589003705305          ESTIMATED ONBOARD DATE               ASUNCION   
253    WG589003705305          ESTIMATED ONBOARD DATE                  CNSHK   
10693  WG589003705305            PLANNED PICK UP DATE         SHENZHEN_CHINA   
10692  WG589003705305    CARGO ARRIVAL AT DESTINATION  ROTTERDAM_NETHERLANDS   
1983   WG589003705305             UPDATE SHIPMENT ETA  BUENAVENTURA_COLOMBIA   
24     WG589003705305  ESTIMATED ARRIVAL TIME TO PORT               HONGKONG   
21     WG589003705305           DAILY TRACK AND TRACE                    NaN   

      EVENT_CONVOLUTION_DATE  
127      2018-12-30 17:00:00  
128      2018-12-31 17:00:00  
126      2019-01-01 17:00:00  
23       2019-01-02 17:00:00  
22       2019-01-03 17:00:00  
253      2019-01-06 19:00:00  
10693    2019-02-08 08:00:00  
10692    2019-02-10 05:24:00  
1983     2019-02-20 22:00:00  
24       2019-03-13 09:00:00  
21       2019-03-13 17:00:00

In [71]:

sample_order = df_event[df_event.EVENT_CODE=='ARRIVAL AT CFS OR CY'].head(1)["loadingOrder"]
df_event[df_event.loadingOrder==sample_order.values[0]].sort_values("EVENT_CONVOLUTION_DATE")

loadingOrder                    EVENT_CODE EVENT_LOCATION_ID  \
59    XG377450862602          ARRIVAL AT CFS OR CY             CNYTN   
5279  XG377450862602  CARGO ARRIVAL AT DESTINATION     ZIPALL_MONACO   
7409  XG377450862602          PLANNED PICK UP DATE    SHENZHEN_CHINA   

     EVENT_CONVOLUTION_DATE  
59      2018-12-29 03:23:00  
5279    2019-01-22 18:00:00  
7409    2019-02-07 03:12:00

In [72]:

sample_order = df_event[df_event.EVENT_CODE=='TRANSIT PORT ATD'].head(1)["loadingOrder"]
df_event[df_event.loadingOrder==sample_order.values[0]].sort_values("EVENT_CONVOLUTION_DATE")

loadingOrder                      EVENT_CODE EVENT_LOCATION_ID  \
97    NP923683478958                TRANSIT PORT ATD  PORT KELANG WEST   
2387  NP923683478958            ARRIVAL AT CFS OR CY             CNYTN   
2547  NP923683478958  ESTIMATED ARRIVAL TIME TO PORT           JAKARTA   
7994  NP923683478958           SHIPMENT ONBOARD DATE         HONG KONG   
7993  NP923683478958                TRANSIT PORT ATA             AEJEA   

     EVENT_CONVOLUTION_DATE  
97      2019-01-01 17:00:00  
2387    2019-01-12 13:21:00  
2547    2019-01-24 17:00:00  
7994    2019-01-25 09:00:00  
7993    2019-01-27 03:00:00

In [73]:

sample_order = df_event[df_event.EVENT_CODE=='IMP CUSTOMS CLEARANCE START'].head(1)["loadingOrder"]
df_event[df_event.loadingOrder==sample_order.values[0]].sort_values("EVENT_CONVOLUTION_DATE")

loadingOrder                   EVENT_CODE EVENT_LOCATION_ID  \
111  FB602851083489        SHIPMENT ONBOARD DATE           YANTIAN   
241  FB602851083489  IMP CUSTOMS CLEARANCE START               RTM   
372  FB602851083489        DAILY TRACK AND TRACE    SHENZHEN_CHINA   

    EVENT_CONVOLUTION_DATE  
111    2019-01-01 17:00:00  
241    2019-01-03 20:00:00  
372    2019-01-10 12:00:00

In [74]:

sample_order = df_event[df_event.EVENT_CODE=='IMP CUSTOMS CLEARANCE FINISHED'].head(1)["loadingOrder"]
df_event[df_event.loadingOrder==sample_order.values[0]].sort_values("EVENT_CONVOLUTION_DATE")

loadingOrder                      EVENT_CODE EVENT_LOCATION_ID  \
377  BM170438548354  IMP CUSTOMS CLEARANCE FINISHED               RTM   

    EVENT_CONVOLUTION_DATE  
377    2019-01-04 20:00:00

In [75]:

sample_order = df_event[df_event.EVENT_CODE=='DISCHARGED'].head(1)["loadingOrder"]
df_event[df_event.loadingOrder==sample_order.values[0]].sort_values("EVENT_CONVOLUTION_DATE")

loadingOrder                      EVENT_CODE    EVENT_LOCATION_ID  \
2494  MT699113559772            PLANNED PICK UP DATE       SHENZHEN_CHINA   
2734  MT699113559772                      DISCHARGED       PORT SAID EAST   
2742  MT699113559772    CARGO ARRIVAL AT DESTINATION     SYDNEY_AUSTRALIA   
2416  MT699113559772                 ARRIVAL AT PORT       QASIM_PAKISTAN   
2496  MT699113559772    CARGO ARRIVAL AT DESTINATION       BEIRUT_LEBANON   
2414  MT699113559772           SHIPMENT ONBOARD DATE                CNSHK   
2415  MT699113559772           SHIPMENT ONBOARD DATE                CNSHK   
2413  MT699113559772           SHIPMENT ONBOARD DATE                CNSHK   
3365  MT699113559772    CARGO ARRIVAL AT DESTINATION  JAKARTA SEAPORT_IDN   
2498  MT699113559772          ESTIMATED ONBOARD DATE                CNSHK   
2738  MT699113559772             UPDATE SHIPMENT ETA  SINGAPORE_SINGAPORE   
2855  MT699113559772                TRANSIT PORT ETA                THLCH   
2694  MT699113559772  ESTIMATED ARRIVAL TIME TO PORT                MYPKE   
2525  MT699113559772          ESTIMATED ONBOARD DATE                CNYTN   
3753  MT699113559772           DAILY TRACK AND TRACE                  NaN   
6754  MT699113559772           SHIPMENT ONBOARD DATE               SHEKOU   
7016  MT699113559772          ESTIMATED ONBOARD DATE                CNYTN   
3657  MT699113559772             UPDATE SHIPMENT ETA        SANTOS_BRAZIL   
2527  MT699113559772             UPDATE SHIPMENT ETA         COLON_PANAMA   
7017  MT699113559772  ESTIMATED ARRIVAL TIME TO PORT                NLRTM   

     EVENT_CONVOLUTION_DATE  
2494    2018-12-14 14:41:00  
2734    2019-01-08 18:00:00  
2742    2019-01-12 07:00:00  
2416    2019-01-12 15:55:00  
2496    2019-01-12 16:00:00  
2414    2019-01-12 22:57:00  
2415    2019-01-12 22:58:00  
2413    2019-01-13 06:00:00  
3365    2019-01-15 11:00:00  
2498    2019-01-16 23:55:00  
2738    2019-01-21 20:49:00  
2855    2019-01-21 21:38:00  
2694    2019-01-22 11:01:00  
2525    2019-01-22 18:27:00  
3753    2019-01-22 19:06:00  
6754    2019-01-24 09:00:00  
7016    2019-02-04 08:00:00  
3657    2019-02-19 05:19:00  
2527    2019-02-24 11:28:00  
7017    2019-02-28 19:00:00

In [76]:
df_merged = pd.merge(df_sample, df_event, on="loadingOrder")
df_merged.head(10)

loadingOrder carrierName                 timestamp  longitude   latitude  \
0  HM302385710947      OYSCFP 2019-01-01 00:01:44+00:00   1.410158  50.771878   
1  HM302385710947      OYSCFP 2019-01-01 00:06:08+00:00   1.420387  50.788977   
2  HM302385710947      OYSCFP 2019-01-01 00:10:16+00:00   1.429968  50.804840   
3  HM302385710947      OYSCFP 2019-01-01 00:13:21+00:00   1.437205  50.816783   
4  HM302385710947      OYSCFP 2019-01-01 00:17:25+00:00   1.446740  50.832388   
5  HM302385710947      OYSCFP 2019-01-01 00:21:32+00:00   1.456205  50.847913   
6  HM302385710947      OYSCFP 2019-01-01 00:25:44+00:00   1.465760  50.863772   
7  HM302385710947      OYSCFP 2019-01-01 00:30:09+00:00   1.475748  50.880260   
8  HM302385710947      OYSCFP 2019-01-01 00:33:19+00:00   1.482908  50.892587   
9  HM302385710947      OYSCFP 2019-01-01 00:37:28+00:00   1.493078  50.908542   

    vesselMMSI  speed  direction vesselNextport         vesselNextportETA  \
0  F4676005751     27       2030      ROTTERDAM  2019-01-01T00:00:00.000Z   
1  F4676005751     27       2110      ROTTERDAM  2019-01-01T00:00:00.000Z   
2  F4676005751     27       2080      ROTTERDAM  2019-01-01T00:00:00.000Z   
3  F4676005751     27       2130      ROTTERDAM  2019-01-01T00:00:00.000Z   
4  F4676005751     27       2030      ROTTERDAM  2019-01-01T00:00:00.000Z   
5  F4676005751     26       2130      ROTTERDAM  2019-01-01T00:00:00.000Z   
6  F4676005751     26       2130      ROTTERDAM  2019-01-01T00:00:00.000Z   
7  F4676005751     26       2000      ROTTERDAM  2019-01-01T00:00:00.000Z   
8  F4676005751     27       2010      ROTTERDAM  2019-01-01T00:00:00.000Z   
9  F4676005751     28       2550      ROTTERDAM  2019-01-01T00:00:00.000Z   

             vesselStatus vesselDatasource TRANSPORT_TRACE  \
0  under way using engine      Coastal AIS             NaN   
1  under way using engine      Coastal AIS             NaN   
2  under way using engine      Coastal AIS             NaN   
3  under way using engine      Coastal AIS             NaN   
4  under way using engine      Coastal AIS             NaN   
5  under way using engine      Coastal AIS             NaN   
6  under way using engine      Coastal AIS             NaN   
7  under way using engine      Coastal AIS             NaN   
8  under way using engine      Coastal AIS             NaN   
9  under way using engine      Coastal AIS             NaN   

              EVENT_CODE EVENT_LOCATION_ID EVENT_CONVOLUTION_DATE  
0  SHIPMENT ONBOARD DATE             SOHAR    2019-01-02 17:00:00  
1  SHIPMENT ONBOARD DATE             SOHAR    2019-01-02 17:00:00  
2  SHIPMENT ONBOARD DATE             SOHAR    2019-01-02 17:00:00  
3  SHIPMENT ONBOARD DATE             SOHAR    2019-01-02 17:00:00  
4  SHIPMENT ONBOARD DATE             SOHAR    2019-01-02 17:00:00  
5  SHIPMENT ONBOARD DATE             SOHAR    2019-01-02 17:00:00  
6  SHIPMENT ONBOARD DATE             SOHAR    2019-01-02 17:00:00  
7  SHIPMENT ONBOARD DATE             SOHAR    2019-01-02 17:00:00  
8  SHIPMENT ONBOARD DATE             SOHAR    2019-01-02 17:00:00  
9  SHIPMENT ONBOARD DATE             SOHAR    2019-01-02 17:00:00

### 简化事件记录

In [77]:
df_event_filtered = df_event[df_event.EVENT_CODE.isin(["TRANSIT PORT ATD","SHIPMENT ONBOARD DATE",
                                                       "TRANSIT PORT ATA","ARRIVAL AT PORT","CARGO ARRIVAL AT DESTINATION"])]
print(df_event.shape[0],df_event_filtered.shape[0])

158341 54364


In [79]:
df_event_filtered = df_event_filtered.drop_duplicates()
df_event_filtered.shape[0]

52477

In [80]:
df_event_filtered = df_event_filtered[df_event_filtered.EVENT_LOCATION_ID.isin(set(df_port.TRANS_NODE_NAME))]

In [81]:
df_event_filtered.to_csv("data/event_fixed_0613.csv",index=False)

### 清理事件间的内部逻辑

- 如果有连续多个相同的事件代码，发生在相同的地点，取时间最早的那个。示例97391，100086，102856
- 如果有连续多个相同的事件代码，发生在不同的地点，时间有很相近，取地理位置合理的那个
- [ ] 在event表中，把trans_code转换成地理位置坐标。根据时间与坐标变化之间的合理性，来判断是否要保留或删除

In [86]:
df_event_sorted = df_event_filtered.groupby("loadingOrder").apply(lambda x: x.sort_values("EVENT_CONVOLUTION_DATE"))

In [89]:
df_event_sorted.to_csv("data/event_fixed_0613.csv",index=False)

In [83]:
df_port[df_port.TRANS_NODE_NAME=="MYPKG"]

TRANS_NODE_NAME   LONGITUDE  LATITUDE   COUNTRY     STATE  \
42             MYPKG  101.361204  3.034709  Malaysia  Selangor   
2438           MYPKG  101.392830  2.999852  Malaysia  Selangor   

                 CITY           REGION  \
42    Pelabuhan Klang  Pelabuhan Utara   
2438            Klang       Kawasan 13   

                                                ADDRESS  PORT_CODE  \
42    Jalan Kerisi, Pelabuhan Utara, 42000 Pelabuhan...        NaN   
2438  1, Persiaran Raja Muda Musa, Kawasan 13, 41100...        NaN   

      TRANSPORT_NODE_ID  
42             636000.0  
2438        272214000.0

In [84]:
df_port[df_port.TRANS_NODE_NAME=="KELANG_MALAYSIA"]

TRANS_NODE_NAME  LONGITUDE  LATITUDE   COUNTRY STATE    CITY REGION  \
1487  KELANG_MALAYSIA   0.831116 -0.386592  MALAYSIA   NaN  KELANG    NaN   

     ADDRESS  PORT_CODE  TRANSPORT_NODE_ID  
1487     NaN        NaN         19283000.0

In [85]:
len(df_event.EVENT_LOCATION_ID.unique())

1050

In [4]:
import pandas as pd
from tqdm import tqdm
import numpy as np
import math

from sklearn.metrics import mean_squared_error,explained_variance_score
from sklearn.model_selection import KFold
import lightgbm as lgb

import os
import glob
import warnings
warnings.filterwarnings('ignore')

test_data_path = 'data/B_testData0626.csv'
test_data = pd.read_csv('F:/grad/HWC/ChuSai/'+test_data_path)
course_list = list(test_data["TRANSPORT_TRACE"].unique())

In [5]:
for i in range(len(course_list)-1,len(course_list)):
    print('==========',course_list[i],'started training   ==========')
    train_set = pd.read_csv('F:/grad/HWC/ChuSai/data/feature/train/'+course_list[i]+'.csv')
    test_set = pd.read_csv('F:/grad/HWC/ChuSai/data/feature/test/'+course_list[i]+'.csv')
    
    train_set['longitude'] = train_set['start_longitude']+train_set['long_diff_from_start']
    test_set['longitude'] = test_set['start_longitude']+test_set['long_diff_from_start']
    
    print('test_set locations =',(test_set['longitude'].min(),test_set['longitude'].max()))
    if test_set['longitude'].max()-test_set['longitude'].min() > 350:
        train_set['longitude'][train_set['longitude']<0] = train_set['longitude'] + 360
        test_set['longitude'][test_set['longitude']<0] = test_set['longitude'] + 360
    train_set = train_set[(train_set['longitude']<test_set['longitude'].max()) & (train_set['longitude']>test_set['longitude'].min())]
    train_set = train_set.drop(['longitude'],axis=1)
    train_set.to_csv('F:/grad/HWC/ChuSai/data/feature/trains/'+course_list[i]+'.csv', index = False)

========== CNSHK-SGSIN-AEJEA started training   ==========
test_set locations = (113.898995, 113.943727)
